### PR TITLE
RMAC-3986: Reload of inner views

### PR DESCRIPTION
**Pull request by tbragaf**
https://outsystemsrd.atlassian.net/browse/RMAC-3986

### CODE REVIEW

File: ReactViewControl/ReactViewRender.cs [hunk 1]
```diff
@@ -201,7 +202,11 @@ internal partial class ReactViewRender : IChildViewHost, IDisposable {
 
             lock (SyncRoot) {
                 var mainFrame = Frames[FrameInfo.MainViewFrameName];
-                Frames.Clear();
+                foreach (var frame in Frames) {
+                    RecoverableFrames[frame.Key] = new WeakReference<FrameInfo>(frame.Value);
```
Comment (joaompneves) [inline ReactViewControl/ReactViewRender.cs:206]: no need to store main frame

File: ReactViewControl/ReactViewRender.cs [hunk 2]
```diff
@@ -201,7 +202,11 @@ internal partial class ReactViewRender : IChildViewHost, IDisposable {
 
             lock (SyncRoot) {
                 var mainFrame = Frames[FrameInfo.MainViewFrameName];
-                Frames.Clear();
+                foreach (var frame in Frames) {
```
Comment (joaompneves) [inline ReactViewControl/ReactViewRender.cs:205]: clear RecoverableFrames before

File: ReactViewControl/ReactViewRender.cs [hunk 3]
```diff
@@ -210,7 +215,9 @@ internal partial class ReactViewRender : IChildViewHost, IDisposable {
         }
 
         public void Dispose() {
-            WebView.Dispose();
+            WebView.Dispose();
+            Frames.Clear();
```
Comment (joaompneves) [inline ReactViewControl/ReactViewRender.cs:219]: not needed

File: ReactViewControl/ReactViewRender.cs [hunk 4]
```diff
@@ -588,13 +595,22 @@ internal partial class ReactViewRender : IChildViewHost, IDisposable {
...
-                Frames[frameName] = frame;
-                AddPlugins(PluginsFactory(), frame);
-            }
-            return frame;
+        private FrameInfo GetOrCreateFrame(string frameName) {
+            if (Frames.TryGetValue(frameName, out var frame)) {
+                return frame;
+            }
+
+            if (RecoverableFrames.TryGetValue(frameName, out var weakReferenceFrame) && weakReferenceFrame.TryGetTarget(out var recoverableFrame)) {
```
Comment (joaompneves) [inline ReactViewControl/ReactViewRender.cs:603]:
```
if (RecoverableFrames.TryGetValue(frameName, out var weakReferenceFrame)) {
    RecoverableFrames.Remove(frameName);
    if (weakReferenceFrame.TryGetTarget(out var recoverableFrame)) {
        Frames[frameName] = recoverableFrame;
        return recoverableFrame;
    }
}

**Review by joaompneves — CHANGES_REQUESTED** [review]

File: ReactViewControl/ReactViewRender.cs [hunk 5]
```diff
@@ -588,13 +595,22 @@ internal partial class ReactViewRender : IChildViewHost, IDisposable {
...
+                Frames[frameName] = recoverableFrame;
+                RecoverableFrames.Remove(frameName);
```
Comment (osplm) [inline ReactViewControl/ReactViewRender.cs:605]: Just a question. As we are adding all Frames, is it possible that we end accumulating frames on RecoverableFrames that are never recreated?

Comment (osplm) [reply ReactViewControl/ReactViewRender.cs:605]: Update: Problem solved with suggestions from Neves that saw only now

Comment (joaompneves) [reply ReactViewControl/ReactViewRender.cs:605]: theres a known issue.. we might end up with garbage (entries) on the RecoverableFrames dictionary without value ()once the weak ref goes away)... only the strings (key). We could have a method to periodically clear the dictionary... although the gains might be residual (we're just storing strings).

**Review by joaompneves — COMMENTED** [review]
Missing a test

**Review by marianacapelo — APPROVED** [review]
Should we include this in the version used by WPF?

**Review by marianacapelo — APPROVED** [review]

**Review by osplm — APPROVED** [review]

File: Tests.ReactView/InnerViewTests.cs [hunk 6]
```diff
@@ -15,17 +15,39 @@ public class InnerViewTests : ReactViewTestBase {
...
+#endif
                 TargetView.InnerView.Load();
+
                 var isLoaded = await taskCompletionSource.Task;
 
                 Assert.IsTrue(isLoaded, "Inner view module was not loaded!");
             });
+        }
+
+        [Test(Description = "Tests inner view reload")]
```
Comment (joaompneves) [inline Tests.ReactView/InnerViewTests.cs:32]: ... when webview is reloaded

File: Tests.ReactView/InnerViewTests.cs [hunk 7]
```diff
@@ -15,17 +15,39 @@ public class InnerViewTests : ReactViewTestBase {
...
+                };
+
+                TargetView.InnerView.Load();
+                TargetView.ExecuteMethod("reload");
+
+                var isReloaded = await taskCompletionSource.Task;
+
+                Assert.IsTrue(isReloaded, "Inner view module was not reloaded!");
+            });
         }
```
Comment (joaompneves) [inline Tests.ReactView/InnerViewTests.cs:51]: missing a memory leak test

File: Tests.ReactView/InnerViewTests.cs [hunk 8]
```diff
@@ -15,18 +16,72 @@ public class InnerViewTests : ReactViewTestBase {
...
+
                 var isLoaded = await taskCompletionSource.Task;
 
                 Assert.IsTrue(isLoaded, "Inner view module was not loaded!");
             });
-        }
+        }
+
+        [Test(Description = "Tests inner view reload when web view is reloaded")]
+        public async Task InnerViewIsReloadedWhenWebviewIsReloaded() {
```
Comment (joaompneves) [inline Tests.ReactView/InnerViewTests.cs:34]: InnerViewIsLoadedWhenWebviewIsReloaded

File: Tests.ReactView/TestAppView/TestApp.tsx [hunk 9]
```diff
@@ -18,20 +18,41 @@ interface IChildViews {
...
         this.firstRenderHtml = this.getHtml();
     }
 
+    setShouldRenderInnerView(shouldRenderInnerView: boolean): void {
+        this.setState({ shouldRenderInnerView });
+    }
+
     renderInnerViewContainer() {
-        return this.props.autoShowInnerView ? <ViewFrame<IChildViews> key="test_frame" name="test" className="" loaded={() => this.innerViewLoadedTask.setResult()} /> : null;
+        if (this.props.autoShowInnerView && this.state.shouldRenderInnerView) {
```
Comment (joaompneves) [inline Tests.ReactView/TestAppView/TestApp.tsx:50]: if ( this.state.shouldRenderInnerView) {

File: Tests.ReactView/TestAppView/TestApp.tsx [hunk 10]
```diff
@@ -18,20 +18,41 @@ interface IChildViews {
     test: InnerView;
 }
 
-class App extends React.Component<IAppProperties> {
+interface IAppState {
+    shouldRenderInnerView: boolean;
+}
+
+interface IAppBehaviors {
```
Comment (joaompneves) [inline Tests.ReactView/TestAppView/TestApp.tsx:25]: not needed

File: Tests.ReactView/TestAppView/TestApp.tsx [hunk 11]
```diff
@@ -18,20 +18,41 @@ interface IChildViews {
...
+        this.setState({ shouldRenderInnerView });
+    }
+
     renderInnerViewContainer() {
-        return this.props.autoShowInnerView ? <ViewFrame<IChildViews> key="test_frame" name="test" className="" loaded={() => this.innerViewLoadedTask.setResult()} /> : null;
+        if (this.props.autoShowInnerView && this.state.shouldRenderInnerView) {
+            return <ViewFrame<IChildViews> key="test_frame" name="test" className="" loaded={() => this.innerViewLoadedTask.setResult()} />;
+        }
+
+        this.props.event("NoInnerView");
```
Comment (joaompneves) [inline Tests.ReactView/TestAppView/TestApp.tsx:54]: dont launch event here .. this is a render method ... use component did mount or update

File: Tests.ReactView/InnerViewTests.cs [hunk 12]
```diff
@@ -15,18 +16,72 @@ public class InnerViewTests : ReactViewTestBase {
...
+                TargetView.InnerView.Loaded += () => {
+                    taskCompletionSource.TrySetResult(true);
+                };
+
+                TargetView.InnerView.Load();
+                await taskCompletionSource.Task;
+
+                // Reload view
+                var targetViewCompletionSource = new TaskCompletionSource<bool>();
+                TargetView.ExecuteMethod("setShouldRenderInnerView", false);
```
Comment (joaompneves) [inline Tests.ReactView/InnerViewTests.cs:70]: TargetView.ExecuteMethod("setShouldRenderInnerView", false); this does not make sense... the state will go away with reload

**Review by joaompneves — CHANGES_REQUESTED** [review]

File: Tests.ReactView/InnerViewTests.cs (hunk 12 above)
Comment (tbragaf) [reply Tests.ReactView/InnerViewTests.cs]: true... but I can't make it work by simply changing the props... :/

Comment (joaompneves) [reply Tests.ReactView/InnerViewTests.cs]: i'm not confident that this is really working...

**Review by joaompneves — APPROVED** [review]

**Review by joaompneves — APPROVED** [review]

**Pull request merged**